Устанавливаем HMMER

In [ ]:
%%bash
wget http://eddylab.org/software/hmmer/hmmer-3.3.2.tar.gz
tar xf hmmer-3.3.2.tar.gz
cd hmmer-3.3.2

--2026-06-10 09:54:16--  http://eddylab.org/software/hmmer/hmmer-3.3.2.tar.gz
Resolving eddylab.org (eddylab.org)... 96.126.110.11, 2600:3c03::f03c:91ff:fec8:383c
Connecting to eddylab.org (eddylab.org)|96.126.110.11|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18213049 (17M) [application/x-gzip]
Saving to: ‘hmmer-3.3.2.tar.gz’

     0K .......... .......... .......... .......... ..........  0%  390K 45s
    50K .......... .......... .......... .......... ..........  0%  778K 34s
   100K .......... .......... .......... .......... ..........  0% 94.2M 23s
   150K .......... .......... .......... .......... ..........  1%  787K 23s
   200K .......... .......... .......... .......... ..........  1%  148M 18s
   250K .......... .......... .......... .......... ..........  1% 38.7M 15s
   300K .......... .......... .......... .......... ..........  1%  145M 13s
   350K .......... .......... .......... .......... ..........  2%  350M 11s
   400K .......... ......

In [ ]:
%%bash
cd hmmer-3.3.2
./configure
make
make install

configure: Configuring HMMER3 for your system.
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking whether to compile using MPI... no
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for gcc option to accept ISO C99... none needed
checking for gcc option to accept ISO Standard C... (cached) none needed
checking how to run the C preprocessor... gcc -E
checking for a BSD-compatible install... /usr/bin/install -c
checking for strip... strip
checking for ranlib... ranlib
checking for ar... /usr/bin/ar
checking whether ln -s works... yes
checking for a sed that does not truncat

In file included from hmmbuild.c:13:
In function ‘usual_master’,
    inlined from ‘main’ at hmmbuild.c:463:7:
../easel/easel.h:164:17: warning: argument 1 range [18446744022169944064, 18446744073709551592] exceeds maximum object size 9223372036854775807 [-Walloc-size-larger-than=]
  164 |     if ( ((p) = malloc(size)) == NULL)  { \
      |                 ^~~~~~~~~~~~
hmmbuild.c:561:3: note: in expansion of macro ‘ESL_ALLOC’
  561 |   ESL_ALLOC(info, sizeof(*info) * infocnt);
      |   ^~~~~~~~~
In file included from hmmbuild.c:6:
hmmbuild.c: In function ‘main’:
/usr/include/stdlib.h:540:14: note: in a call to allocation function ‘malloc’ declared here
  540 | extern void *malloc (size_t __size) __THROW __attribute_malloc__
      |              ^~~~~~
In file included from hmmscan.c:14:
In function ‘serial_master’,
    inlined from ‘main’ at hmmscan.c:302:16:
../easel/easel.h:164:17: warning: argument 1 range [18446743987810205696, 18446744073709551576] exceeds maximum object size 9223

Скачиваем Pfam


In [ ]:
#!wget --spider http://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam25.0/Pfam-A.full.gz

In [ ]:
#!wget ftp://ftp.sanger.ac.uk/pub/databases/Pfam/releases/Pfam25.0/Pfam-A.full.gz.

In [ ]:
!wget http://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam35.0/Pfam-A.hmm.gz

--2026-06-10 10:02:56--  http://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam35.0/Pfam-A.hmm.gz
Resolving ftp.ebi.ac.uk (ftp.ebi.ac.uk)... 193.62.193.165
Connecting to ftp.ebi.ac.uk (ftp.ebi.ac.uk)|193.62.193.165|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 293000230 (279M) [application/x-gzip]
Saving to: ‘Pfam-A.hmm.gz.1’

Pfam-A.hmm.gz.1       7%[>                   ]  20.07M   839KB/s    eta 5m 25s ^C


In [ ]:
!wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/001/194/135/GCF_001194135.2_ASM119413v2/GCF_001194135.2_ASM119413v2_genomic.fna.gz

--2026-06-10 10:03:24--  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/001/194/135/GCF_001194135.2_ASM119413v2/GCF_001194135.2_ASM119413v2_genomic.fna.gz
           => ‘GCF_001194135.2_ASM119413v2_genomic.fna.gz’
Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.10, 130.14.250.11, 2607:f220:41e:250::11, ...
Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.10|:21... connected.
Logging in as anonymous ... Logged in!
==> SYST ... done.    ==> PWD ... done.
==> TYPE I ... done.  ==> CWD (1) /genomes/all/GCF/001/194/135/GCF_001194135.2_ASM119413v2 ... done.
==> SIZE GCF_001194135.2_ASM119413v2_genomic.fna.gz ... 643198082
==> PASV ... done.    ==> RETR GCF_001194135.2_ASM119413v2_genomic.fna.gz ... done.
Length: 643198082 (613M) (unauthoritative)

GCF_001194135.2_ASM 100%[===================>] 613.40M  79.0MB/s    in 8.4s    

2026-06-10 10:03:34 (72.6 MB/s) - ‘GCF_001194135.2_ASM119413v2_genomic.fna.gz’ saved [643198082]



In [ ]:
!wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/001/194/135/GCF_001194135.2_ASM119413v2/GCF_001194135.2_ASM119413v2_genomic.gtf.gz
!gzip -d GCF*

--2026-05-26 16:33:56--  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/001/194/135/GCF_001194135.2_ASM119413v2/GCF_001194135.2_ASM119413v2_genomic.gtf.gz
           => ‘GCF_001194135.2_ASM119413v2_genomic.gtf.gz’
Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.11, 130.14.250.12, 2607:f220:41e:250::12, ...
Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.11|:21... connected.
Logging in as anonymous ... Logged in!
==> SYST ... done.    ==> PWD ... done.
==> TYPE I ... done.  ==> CWD (1) /genomes/all/GCF/001/194/135/GCF_001194135.2_ASM119413v2 ... done.
==> SIZE GCF_001194135.2_ASM119413v2_genomic.gtf.gz ... 7963480
==> PASV ... done.    ==> RETR GCF_001194135.2_ASM119413v2_genomic.gtf.gz ... done.
Length: 7963480 (7.6M) (unauthoritative)

GCF_001194135.2_ASM 100%[===================>]   7.59M  29.5MB/s    in 0.3s    

2026-05-26 16:33:56 (29.5 MB/s) - ‘GCF_001194135.2_ASM119413v2_genomic.gtf.gz’ saved [7963480]

gzip: GCF_001194135.2_ASM119413v2_featu

Извлекаем из архива

In [ ]:
!gunzip Pfam-A.hmm.gz

^C


Создаем нужные для hmmer'a файлы

In [ ]:
!hmmpress Pfam-A.hmm


Error: File existence/permissions problem in trying to open HMM file Pfam-A.hmm.
HMM file Pfam-A.hmm not found (nor an .h3m binary of it)



In [ ]:
!head -n 40 Pfam-A.hmm

HMMER3/f [3.1b2 | February 2015]
NAME  1-cysPrx_C
ACC   PF10417.12
DESC  C-terminal domain of 1-Cys peroxiredoxin
LENG  40
ALPH  amino
RF    no
MM    no
CONS  yes
CS    yes
MAP   yes
DATE  Thu Nov  4 19:20:02 2021
NSEQ  40
EFFN  17.426758
CKSUM 4086680297
GA    21.1 21.1;
TC    21.1 21.1;
NC    21 21;
BM    hmmbuild HMM.ann SEED.ann
SM    hmmsearch -Z 61295632 -E 1000 --cpu 4 HMM pfamseq
STATS LOCAL MSV       -7.5463  0.71948
STATS LOCAL VITERBI   -7.8624  0.71948
STATS LOCAL FORWARD   -4.3303  0.71948
HMM          A        C        D        E        F        G        H        I        K        L        M        N        P        Q        R        S        T        V        W        Y   
            m->m     m->i     m->d     i->m     i->i     d->m     d->d
  COMPO   2.28046  4.31208  2.83393  2.63913  3.90855  2.69988  3.89812  3.33401  2.56310  2.85023  3.99954  3.22924  2.52123  2.90328  3.31238  2.94055  2.70512  2.59551  3.49266  3.82715
          2.68618  4.42225  2.77519  2.7312

Смотрим статистику

In [ ]:
!hmmstat Pfam-A.hmm.h3m | head -n 20

# hmmstat :: display summary statistics for a profile file
# HMMER 3.3.2 (Nov 2020); http://hmmer.org/
# Copyright (C) 2020 Howard Hughes Medical Institute.
# Freely distributed under the BSD open source license.
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
#
# idx  name                 accession        nseq eff_nseq      M relent   info p relE compKL
# ---- -------------------- ------------ -------- -------- ------ ------ ------ ------ ------
1      1-cysPrx_C           PF10417.12         40    17.43     40   1.37   1.32   1.29   0.09
2      120_Rick_ant         PF12574.11          2     0.45    238   0.59   0.60   0.52   0.02
3      12TM_1               PF09847.12          7     2.28    449   0.59   0.62   0.51   0.09
4      14-3-3               PF00244.23        149     2.50    222   0.59   0.59   0.52   0.05
5      17kDa_Anti_2         PF16998.8           6     0.99    116   0.59   0.57   0.52   0.03
6      2-Hacid_dh           PF00389.33         78    

The columns are:

**idx**
The index of this profile, numbering each on in the file starting from 1.

**name**
The name of the profile.

**accession**
The optional accession of the profile, or "-" if there is none.

**nseq**
The number of sequences that the profile was estimated from.

**eff_nseq**
The effective number of sequences that the profile was estimated from, after HMMER applied an effective sequence number calculation such as the default entropy weighting.

**M**
The length of the model in consensus residues (match states).

**relent**
Mean relative entropy per match state, in bits. This is the expected (mean) score per consensus position. This is what the default entropy-weighting method for effective sequence number estimation focuses on, so for default HMMER3 models, you expect this value to reflect the default target for entropy-weighting.

**info**
Mean information content per match state, in bits. Probably not useful. Information content is a slightly different calculation than relative entropy.

**p relE**
Mean positional relative entropy, in bits. This is a fancier version of the per-match-state relative entropy, taking into account the transition (insertion/deletion) probabilities; it may be a more accurate estimation of the average score contributed per model consensus position.

**compKL**
Kullback-Leibler distance between the model's overall average residue composition and the default background frequency distribution. The higher this number, the more biased the residue composition of the profile is. Highly biased profiles can slow the HMMER3 acceleration pipeline, by causing too many nonhomologous sequences to pass the filters.

Индексируем файл

In [ ]:
!hmmfetch --index Pfam-A.hmm.h3m


Error: File existence/permissions problem in trying to open HMM file Pfam-A.hmm.h3m.
HMM file Pfam-A.hmm.h3m not found (nor an .h3m binary of it)



Извлекаем информацию о нужном нам домене

In [ ]:
!hmmfetch Pfam-A.hmm.h3m  PF02295.20 > Z-alpha.hmm

In [ ]:
!hmmfetch Pfam-A.hmm.h3m PF00022 > Actin.hmm


Error: HMM PF00022 not found in SSI index for file Pfam-A.hmm.h3m



Обратите внимание на наличие версии в ключе

In [ ]:
!grep PF04824 Pfam-A.hmm

ACC   PF04824.19


In [ ]:
!cat Z-alpha.hmm

HMMER3/f [3.3.2 | Nov 2020]
NAME  z-alpha
ACC   PF02295.20
DESC  Adenosine deaminase z-alpha domain
LENG  67
ALPH  amino
RF    no
MM    no
CONS  yes
CS    yes
MAP   yes
DATE  Wed Oct 20 16:10:49 2021
NSEQ  5
EFFN  2.075195
CKSUM 1478172488
GA    21.00 21.00
TC    21.00 21.00
NC    20.90 20.90
STATS LOCAL MSV       -8.4752  0.71882
STATS LOCAL VITERBI   -8.9866  0.71882
STATS LOCAL FORWARD   -4.1378  0.71882
HMM          A        C        D        E        F        G        H        I        K        L        M        N        P        Q        R        S        T        V        W        Y   
            m->m     m->i     m->d     i->m     i->i     d->m     d->d
  COMPO   2.58911  4.21516  2.84992  2.50534  3.88984  3.12937  3.91130  2.82147  2.48816  2.28753  3.65784  2.98103  3.07591  3.05289  2.86388  2.80780  2.95498  2.78660  4.09696  3.70005
          2.68618  4.42225  2.77519  2.73123  3.46354  2.40513  3.72494  3.29354  2.67741  2.69355  4.24690  2.90347  2.73739  3.18146  2.89

In [ ]:
!head -4 /content/GCF_001194135.2_ASM119413v2_protein.faa

>NP_001310715.1 sonic hedgehog protein-like precursor [Octopus bimaculoides]
MPQRSLRHQLGMILVFFLLVTSHSLACGPGRGPGKRRGPRKRTPLVFKQHIPNVSENTVGASGIHEGKITKPDPRFKEMV
TNLNPNIVFRDEEENNEDRVMSKRCKDKLNTLAIAVMNEWPGVKLRVTEAWDTQGHHAPTSLHYEGRAVDITTSDRVRSR
YGMLARLAVEAGFDWVYYESRSHIHCSVRSDSLDTTHYGGCFPRTGKVVVRNKGTITLDQLKVGDSVLSVDLQGELTYSE


In [ ]:
!hmmstat Z-alpha.hmm

# hmmstat :: display summary statistics for a profile file
# HMMER 3.3.2 (Nov 2020); http://hmmer.org/
# Copyright (C) 2020 Howard Hughes Medical Institute.
# Freely distributed under the BSD open source license.
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
#
# idx  name                 accession        nseq eff_nseq      M relent   info p relE compKL
# ---- -------------------- ------------ -------- -------- ------ ------ ------ ------ ------
1      z-alpha              PF02295.20          5     2.08     67   0.84   0.83   0.79   0.03


In [ ]:
#!hmmsearch Z-alpha.hmm Human_proteome_Chr20.fasta

In [ ]:
!hmmsearch Z-alpha.hmm /content/GCF_001194135.2_ASM119413v2_protein.faa

# hmmsearch :: search profile(s) against a sequence database
# HMMER 3.3.2 (Nov 2020); http://hmmer.org/
# Copyright (C) 2020 Howard Hughes Medical Institute.
# Freely distributed under the BSD open source license.
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# query HMM file:                  Z-alpha.hmm
# target sequence database:        /content/GCF_001194135.2_ASM119413v2_protein.faa
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

Query:       z-alpha  [M=67]
Accession:   PF02295.20
Description: Adenosine deaminase z-alpha domain
Scores for complete sequences (score includes all domains):
   --- full sequence ---   --- best 1 domain ---    -#dom-
    E-value  score  bias    E-value  score  bias    exp  N  Sequence       Description
    ------- ------ -----    ------- ------ -----   ---- --  --------       -----------
    3.6e-07   30.8   0.0    4.8e-06   27.2   0.0    2.4  2  XP_014773876.1  double-stranded RNA-specific adenos

In [ ]:
!hmmsearch -h

# hmmsearch :: search profile(s) against a sequence database
# HMMER 3.3.2 (Nov 2020); http://hmmer.org/
# Copyright (C) 2020 Howard Hughes Medical Institute.
# Freely distributed under the BSD open source license.
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Usage: hmmsearch [options] <hmmfile> <seqdb>

Basic options:
  -h : show brief help on version and usage

Options directing output:
  -o <f>           : direct output to file <f>, not stdout
  -A <f>           : save multiple alignment of all hits to file <f>
  --tblout <f>     : save parseable table of per-sequence hits to file <f>
  --domtblout <f>  : save parseable table of per-domain hits to file <f>
  --pfamtblout <f> : save table of hits and domains to file, in Pfam format <f>
  --acc            : prefer accessions over names in output
  --noali          : don't output alignments, so output is smaller
  --notextw        : unlimit ASCII text output line width
  --textw <n>      : set max width of 

In [ ]:
import pandas as pd

# Загрузить feature table для координат
df = pd.read_csv("/content/GCF_001194135.2_ASM119413v2_feature_table.txt",
                 sep="\t", low_memory=False)

# Найти наши белки
hits = ["XP_014773876.1", "XP_014773875.1", "XP_052824128.1"]

result = df[df["product_accession"].isin(hits)][
    ["# feature", "product_accession", "name", "symbol",
     "genomic_accession", "start", "end", "strand", "chromosome"]
]

print(result.to_string())

      # feature product_accession                                                         name        symbol genomic_accession     start       end strand chromosome
23265       CDS    XP_014773875.1  double-stranded RNA-specific adenosine deaminase isoform X2  LOC106871755       NC_068985.1  18319279  18393671      -          5
23266       CDS    XP_052824128.1  double-stranded RNA-specific adenosine deaminase isoform X1  LOC106871755       NC_068985.1  18319279  18393671      -          5
23267       CDS    XP_014773876.1  double-stranded RNA-specific adenosine deaminase isoform X3  LOC106871755       NC_068985.1  18319279  18363518      -          5


# Проверка на наличие CTCF и когезина


In [ ]:
!grep PF02463  Pfam-A.hmm

grep: Pfam-A.hmm: No such file or directory


In [ ]:
!hmmfetch Pfam-A.hmm.h3m  PF04824.19 > Rad21.hmm


Error: File existence/permissions problem in trying to open HMM file Pfam-A.hmm.h3m.
HMM file Pfam-A.hmm.h3m not found (nor an .h3m binary of it)



In [ ]:
!hmmfetch Pfam-A.hmm.h3m  PF00096.29 > zf.hmm

In [ ]:
!hmmfetch Pfam-A.hmm.h3m  PF02463.22 > smc.hmm

In [ ]:
!hmmsearch --tblout ctcf_hits.txt CTCF_Nterm.hmm /content/GCF_001194135.2_ASM119413v2_protein.faa

# hmmsearch :: search profile(s) against a sequence database
# HMMER 3.3.2 (Nov 2020); http://hmmer.org/
# Copyright (C) 2020 Howard Hughes Medical Institute.
# Freely distributed under the BSD open source license.
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# query HMM file:                  CTCF_Nterm.hmm
# target sequence database:        /content/GCF_001194135.2_ASM119413v2_protein.faa
# per-seq hits tabular output:     ctcf_hits.txt
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

Query:       EFP  [M=59]
Accession:   PF01132.28
Description: Elongation factor P (EF-P) OB domain
Scores for complete sequences (score includes all domains):
   --- full sequence ---   --- best 1 domain ---    -#dom-
    E-value  score  bias    E-value  score  bias    exp  N  Sequence Description
    ------- ------ -----    ------- ------ -----   ---- --  -------- -----------

   [No hits detected that satisfy reporting thresholds]


Domain annotat

In [ ]:
!hmmsearch --tblout ctcf_hits.txt CTCF_Nterm.hmm /content/GCF_001194135.2_ASM119413v2_protein.faa
!hmmsearch --tblout smc_hits.txt   smc.hmm     /content/GCF_001194135.2_ASM119413v2_protein.faa
!hmmsearch --tblout rad21_hits.txt Rad21.hmm   /content/GCF_001194135.2_ASM119413v2_protein.faa

Выходные данные были обрезаны до нескольких последних строк (5000).

  == domain 3  score: 9.4 bits;  conditional E-value: 0.19
                     ETTTTEEESSHHHHHHHHHC.C CS
         zf-C2H2   3 CpdCgksFkrksnLkrHirt.H 23 
                     C++C k F++k+ Lk H    H
  XP_052827096.1 719 CEICRKEFCNKYFLKIHKANmH 740
                     *****************98777 PP

  == domain 4  score: 15.0 bits;  conditional E-value: 0.0033
                     EETTTTEEESSHHHHHHHHHC.C CS
         zf-C2H2   2 kCpdCgksFkrksnLkrHirt.H 23 
                     +C++C++ F++k++++ H++  H
  XP_052827096.1 891 YCNICNREFSSKYSYRIHRMNvH 913
                     6*****************99888 PP

  == domain 5  score: 8.8 bits;  conditional E-value: 0.3
                      EETTTTEEESSHHHHHHHHHC.C CS
         zf-C2H2    2 kCpdCgksFkrksnLkrHirt.H 23  
                      +C++C+k  ++k+ Lk+H  + H
  XP_052827096.1  978 TCDICSKELCNKYFLKNHKLKvH 1000
                      7*****************99988 PP

  == domain 6  score: 5.8 bits

In [ ]:
import pandas as pd

def parse_tblout(filename):
    rows = []
    with open(filename) as f:
        for line in f:
            if line.startswith('#'): continue
            cols = line.split()
            if len(cols) < 10: continue
            rows.append({
                'protein_id': cols[0],
                'E-value':    float(cols[4]),
                'score':      float(cols[5]),
                'description': ' '.join(cols[18:])
            })
    return pd.DataFrame(rows)

threshold = 1e-5

for fname, label in [('ctcf_hits.txt', 'CTCF / zf-C2H2'),
                     ('smc_hits.txt',  'SMC'),
                     ('rad21_hits.txt','RAD21')]:
    df = parse_tblout(fname)
    sig = df[df['E-value'] < threshold].sort_values('E-value')
    print(f"\n=== {label} (E < {threshold}) : {len(sig)} hits ===")
    print(sig.to_string(index=False))

KeyError: 'E-value'

In [ ]:

!wget "https://www.ebi.ac.uk/interpro/wwwapi//entry/pfam/PF01132?annotation=hmm" \
     -O CTCF_Nterm.hmm.gz && gunzip CTCF_Nterm.hmm.gz

--2026-06-10 09:56:36--  https://www.ebi.ac.uk/interpro/wwwapi//entry/pfam/PF01132?annotation=hmm
Resolving www.ebi.ac.uk (www.ebi.ac.uk)... 193.62.193.80
Connecting to www.ebi.ac.uk (www.ebi.ac.uk)|193.62.193.80|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6815 (6.7K) [application/gzip]
Saving to: ‘CTCF_Nterm.hmm.gz’

CTCF_Nterm.hmm.gz   100%[===================>]   6.66K  --.-KB/s    in 0.02s   

2026-06-10 09:56:37 (373 KB/s) - ‘CTCF_Nterm.hmm.gz’ saved [6815/6815]

gzip: CTCF_Nterm.hmm already exists; do you wish to overwrite (y or n)? n
	not overwritten


In [ ]:
!apt-get install -y mafft && \
 mafft --auto ctcf_known.fasta > ctcf_aligned.fasta && \
 hmmbuild ctcf_custom.hmm ctcf_aligned.fasta && \
 hmmsearch -E 1e-5 --tblout ctcf_custom_hits.txt \
           ctcf_custom.hmm /content/GCF_001194135.2_ASM119413v2_protein.faa

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
mafft is already the newest version (7.490-1).
0 upgraded, 0 newly installed, 0 to remove and 3 not upgraded.
/usr/bin/mafft: Cannot open ctcf_known.fasta.

/usr/bin/mafft: line 1233: /tmp/mafft.1kdJyGcan8/infile: No such file or directory
awk: cannot open /tmp/mafft.1kdJyGcan8/size (No such file or directory)
awk: cannot open /tmp/mafft.1kdJyGcan8/size (No such file or directory)
/usr/bin/mafft: line 1237: [: too many arguments
/usr/bin/mafft: line 1242: [: too many arguments
/usr/bin/mafft: line 1247: [: too many arguments
/usr/bin/mafft: line 1252: [: -lt: unary operator expected
/usr/bin/mafft: line 1257: [: -lt: unary operator expected
/usr/bin/mafft: line 1264: [: -lt: unary operator expected
/usr/bin/mafft: line 1271: [: -lt: unary operator expected
------------------------------------------------------------------------------
  MAFFT v7.490 (2021/Oct/30)
  https://mafft.cbrc.jp/alig

**CTCF (профиль zf-C2H2)
CTCF не найден как отдельный ген.**

Все хиты — это просто белки с C2H2 цинковыми пальцами (их у осьминога очень много из-за массовой экспансии ZF-генов).

Ни один хит не имеет характерной архитектуры CTCF — 11 тандемных доменов с нужным spacing. Белок LOC106879997 имеет 21 C2H2 домен, но это не CTCF — скорее всего транскрипционный репрессор другого семейства.


Вывод: CTCF у O. bimaculoides отсутствует — это согласуется с литературой (у большинства моллюсков CTCF утерян или сильно редуцирован).

SMC — 35 хитов.

RAD21 — 1 хит (XP_052823579.1, E=2.4e-10). Единственный гомолог клеспина/RAD21 — ключевой субъединицы когезина.

Поскольку CTCF отсутствует, петли хроматина у осьминога организованы без CTCF-зависимого механизма. Когезин при этом есть — значит петли могут формироваться через другой механизм (экструзия когезином без CTCF-якорей, как показано у некоторых беспозвоночных)

Топ-найденных генов:

1.  LOC106877206 - F-box DNA helicase 1 (Ремоделирование хроматина)
2.  LOC106876744 - GATA zinc finger protein 16 (Транскрипционный фактор, эпигенетическая регуляция)
3. LOC106880212 ATF-4 (Ключевой транскрипционный фактор стресс-ответа)
4. LOC106877680 - REC8 homolog (Мейотический когезин)
5. LOC106883024 - histone H2A-like

In [ ]:
!pip install biopython

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 41.4 MB/s eta 0:00:00


Мы выбрали следующие эпигенетические гены:

In [ ]:
from Bio import SeqIO

targets = {
    "DNMT1":  ["cytosine-5)-methyltransferase 1"],
    "TET3":   ["methylcytosine dioxygenase tet3"],
    "EZH2":   ["enhancer of zeste homolog 2"],
    "KMT2A":  ["lysine N-methyltransferase 2A", "myeloid/lymphoid"],
    "CHD1":   ["chromodomain-helicase-DNA-binding protein 1"],
    "BRD3":   ["bromodomain-containing protein 3"],
    "ADAR1":  ["RNA-specific adenosine deaminase isoform X1"],
    "UHRF1":  ["ubiquitin-protein ligase UHRF1", "E3 ubiquitin-protein ligase UHRF1"],
}

found = {}

for record in SeqIO.parse("GCF_001194135.2_ASM119413v2_protein.faa", "fasta"):
    full_text = (record.id + " " + record.description).lower()
    for gene_name, keywords in targets.items():
        if gene_name not in found:
            if any(kw.lower() in full_text for kw in keywords):
                found[gene_name] = record
                print(f"✓ {gene_name}: {record.id} — {record.description[:60]}")

print(f"\nНайдено: {len(found)}/{len(targets)}")


with open("block1_epigenetic_8genes_obimaculoides.fasta", "w") as out:
    SeqIO.write(list(found.values()), out, "fasta")

✓ TET3: XP_014772827.1 — XP_014772827.1 methylcytosine dioxygenase tet3 [Octopus bima
✓ UHRF1: XP_014774607.1 — XP_014774607.1 E3 ubiquitin-protein ligase UHRF1 [Octopus bi
✓ CHD1: XP_014777205.1 — XP_014777205.1 chromodomain-helicase-DNA-binding protein 1-l
✓ DNMT1: XP_014781622.1 — XP_014781622.1 DNA (cytosine-5)-methyltransferase 1 [Octopus
✓ KMT2A: XP_014789911.2 — XP_014789911.2 histone-lysine N-methyltransferase 2A [Octopu
✓ BRD3: XP_052821656.1 — XP_052821656.1 bromodomain-containing protein 3 isoform X1 [
✓ ADAR1: XP_052824128.1 — XP_052824128.1 double-stranded RNA-specific adenosine deamin

Найдено: 7/8


# Анализ генов, в промоторах которых есть Z-DNA



1.   LOC106877206 - F-box DNA helicase 1
2. LOC106869367 - double-stranded RNA-specific editase 1 (ADAR1)
3.  LOC106876744 - GATA zinc finger protein 16
4.  LOC106880212 - ATF-4
5.  LOC106877680 - REC8 homolog
6.  LOC106883024 - histone H2A-like

In [ ]:
import pandas as pd
from collections import defaultdict

rows = []
with open('/content/ZDNA_promoters-2.csv') as f:
    for line in f:
        cols = line.strip().split('\t')
        if len(cols) < 11: continue
        try:
            score = float(cols[3])
            gene = cols[7]
            desc = cols[10]
            rows.append((score, gene, desc))
        except:
            continue


gene_max = defaultdict(float)
gene_desc = {}
for score, gene, desc in rows:
    if score > gene_max[gene]:
        gene_max[gene] = score
        gene_desc[gene] = desc

keywords = ['transcription', 'zinc finger', 'histone', 'chromatin',
            'kinase', 'homeobox', 'factor', 'helicase', 'replication',
            'repair', 'recombination', 'cyclin', 'cohesin', 'SMC']

filtered = [(score, gene, gene_desc[gene])
            for gene, score in gene_max.items()
            if any(kw.lower() in gene_desc[gene].lower() for kw in keywords)
            and 'uncharacterized' not in gene_desc[gene].lower()
            and 'transfer RNA' not in gene_desc[gene]]

filtered.sort(reverse=True)
for s, g, d in filtered[:15]:
    print(f'{s:.3f}  {g}  {d}')

0.978  LOC106877206  F-box DNA helicase 1
0.973  LOC106875300  probable serine/threonine-protein kinase tsuA
0.973  LOC106875450  guided entry of tail-anchored proteins factor 1
0.971  LOC106876744  GATA zinc finger domain-containing protein 16
0.971  LOC106876012  ATP synthase mitochondrial F1 complex assembly factor 1
0.963  LOC106880212  cyclic AMP-dependent transcription factor ATF-4
0.963  LOC106877680  meiotic recombination protein REC8 homolog
0.957  LOC106874904  zinc finger protein 345
0.957  LOC106883024  histone H2A-like
0.957  LOC106873685  SRSF protein kinase 2
0.956  LOC106875312  5'-AMP-activated protein kinase catalytic subunit alpha-2
0.955  LOC106879507  nuclear transcription factor Y subunit gamma
0.953  LOC106868314  probable serine/threonine-protein kinase DDB_G0267686
0.953  LOC106880956  zinc finger protein 239-like
0.953  LOC106880944  zinc finger protein 271-like


In [ ]:
genes = [
    "LOC106877206", "LOC106876744", "LOC106880212",
    "LOC106877680", "LOC106883024", "LOC106869367",
]

rows = []
with open('/content/ZDNA_promoters-2.csv') as f:
    for line in f:
        cols = line.strip().split('\t')
        if len(cols) < 11: continue
        if cols[7] in genes:
            rows.append(cols)

import pandas as pd
df = pd.DataFrame(rows, columns=['chr','zdna_start','zdna_end','score',
                                  'chr2','prom_start','prom_end','gene',
                                  'dot','strand','description'])
print(df[['gene','chr','prom_start','prom_end','strand','description']].drop_duplicates('gene').to_string(index=False))

        gene         chr prom_start  prom_end strand                                     description
LOC106876744 NC_068985.1  111438431 111440430      +   GATA zinc finger domain-containing protein 16
LOC106869367 NC_068986.1    8701764   8703763      +          double-stranded RNA-specific editase 1
LOC106877206 NC_068986.1   66535076  66537075      -                            F-box DNA helicase 1
LOC106880212 NC_068987.1   96830025  96832024      - cyclic AMP-dependent transcription factor ATF-4
LOC106883024 NC_068988.1   39193962  39195961      +                                histone H2A-like
LOC106877680 NC_068994.1   12891038  12893037      +      meiotic recombination protein REC8 homolog


In [ ]:

bed_content = """NC_068985.1\t111438431\t111440430\tLOC106876744_GATA_ZF\t0\t+
NC_068986.1\t8701764\t8703763\tLOC106869367_ADAR\t0\t+
NC_068986.1\t66535076\t66537075\tLOC106877206_Helicase\t0\t-
NC_068987.1\t96830025\t96832024\tLOC106880212_ATF4\t0\t-
NC_068988.1\t39193962\t39195961\tLOC106883024_H2A\t0\t+
NC_068994.1\t12891038\t12893037\tLOC106877680_REC8\t0\t+"""

with open('/content/zdna_promoters.bed', 'w') as f:
    f.write(bed_content)
print("BED файл создан")

BED файл создан


In [ ]:
!samtools faidx /content/GCF_001194135.2_ASM119413v2_genomic.fna


!apt-get install -y bedtools
!bedtools getfasta -fi octopus_genome.fna \
                  -bed /content/zdna_promoters.bed \
                  -s -name \
                  -fo zdna_promoters_sequences.fasta

!cat zdna_promoters_sequences.fasta

# Анализ генов, в промоторах которых есть G-квадруплексы

1.   LOC106870414 - tropomyosin
2.   LOC106878893 - Mediator of RNA Pol II subunit 28
3.   LOC106872502 - SIK1 kinase
4.   LOC106876456 - homeobox abdominal-A
5.   LOC106882331zinc - finger protein 271




In [ ]:
genes = ["LOC106870414", "LOC106878893", "LOC106872502", "LOC106876456", "LOC106882331"]

with open('/content/pqs in promoters.tsv') as f:
    next(f)
    for line in f:
        cols = line.strip().split('\t')
        if cols[3] in genes:
            print('\t'.join(cols))

chr5	59979221	59979256	LOC106872502	+	serine/threonine-protein kinase SIK1
chr5	59979292	59979315	LOC106872502	+	serine/threonine-protein kinase SIK1
chr5	63314887	63314986	LOC106870414	+	tropomyosin Tod p 1.0102
chr5	63315038	63315082	LOC106870414	+	tropomyosin Tod p 1.0102
chr7	104666089	104666117	LOC106876456	+	homeobox protein abdominal-A
chr7	104666595	104666621	LOC106876456	+	homeobox protein abdominal-A
chr8	5465835	5465874	LOC106878893	+	mediator of RNA polymerase II transcription subunit 28
chr8	5465933	5465953	LOC106878893	+	mediator of RNA polymerase II transcription subunit 28
chr10	6217647	6217676	LOC106882331	+	zinc finger protein 271
chr10	6217699	6217720	LOC106882331	+	zinc finger protein 271
